bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -256,10 +256,12 @@ public RackawareEnsemblePlacementPolicyImpl initialize(ClientConfiguration conf,
         DNSToSwitchMapping dnsResolver;
         if (optionalDnsResolver.isPresent()) {
             dnsResolver = optionalDnsResolver.get();
+            dnsResolver.setBookieAddressResolver(bookieAddressResolver);
```

Comment:
I'd assume that provided `optionalDnsResolver` should be ready to use / have the `bookieAddressResolver` set, as it happens there:
https://github.com/apache/bookkeeper/blob/97818f5123999396e66f5246420d3c7e3d25f53d/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeper.java#L500-L507

Have you traced why we happen to have the optional dns resolver provided here yet not properly configured?
I think the right place to fix the issue is where the dns resolver is created and passed to the dnsResolver.
Shall we move bookieAddressResolver into  DNSToSwitchMapping constructor (or Builder?) and make it mandatory (preconditions check)?

Reply:
Agree

Reply:
Because here it missed condition that dnsResolver is null, so bookieAddressResolver would not be set

Reply:
@gaozhangmin `RackawareEnsemblePlacementPolicyImpl` gets Optional (stack from `initializeEnsemblePlacementPolicy` down to the policy will do `Optional.ofNullable(dnsResolver)` to pass it.) 
`Optional.isPresent()` in `RackawareEnsemblePlacementPolicyImpl` would be false in case of null (which is handled).

Where is the dnsResolver created and set, how did it end up partially configured?

Reply:
if dnsResolver is null, `RackawareEnsemblePlacementPolicyImpl.java ` will create a Configurable resolver. here, bookieAddressResolver should be set. or ,it will throw NPE

I checked the code,  `ZoneawareEnsemblePlacementPolicyImpl` also has the same problem, i will fix it.

Reply:
Configure dnsResolver where it is created/before passing it to downstream components.
This may be a better way fixing the issue.  But I am not sure it will be accepted, cause too many code changes,

Reply:
@gaozhangmin let me reiterate my point.
dns resolver is either null or not in this constructor https://github.com/apache/bookkeeper/blob/c7236adc3cb659e65ae5ce53b7156569d7f50ebd/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeper.java#L410 and it is used to initialize placement policy there. In case it is null the policy will create and initialize (correctly, with address resolver) a new one. If it is not null, the constructor will set the resolver.

The change below, where the policy creates the new dnsResolver, now sets the address resolver is absolutely fine and a fix I am agreeing with.
What I do not understand is why do set the address resolver in case above:
```
        if (optionalDnsResolver.isPresent()) {
            dnsResolver = optionalDnsResolver.get();

  // Why is the next line needed? the policy have to have address resolver set already
  // 
            dnsResolver.setBookieAddressResolver(bookieAddressResolver);
```

The BK client constructor should not be used directly (outside of unit tests) and the builder should be used.
For example:
- here the dns resolver is passed to the builder: https://github.com/apache/bookkeeper/blob/c7236adc3cb659e65ae5ce53b7156569d7f50ebd/stream/distributedlog/core/src/main/java/org/apache/distributedlog/BookKeeperClient.java#L110-L125 and I'd expect it to get configured with the address resolver.
- here the dns resolver is not passed to the builder https://github.com/apache/bookkeeper/blob/9f06a4ce82f61a9dc4b086650a301734332f4246/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L270 and I expect placement policy to create and iniialize a new one.

So, either the part is I am confused about is not actually needed or I am missing specific scenario, but in that case it needs to be addressed differently.

Reply:
Remove duplicate setBookieAddressResolver when optionalDnsResolver isPresent